contracts/src/token/erc20/utils/safe_erc20.rs

```diff
@@ -0,0 +1,229 @@
+        let current_allowance =
+            erc20.allowance(call, address(), spender).or({
+                Err(Error::SafeErc20FailedOperation(SafeErc20FailedOperation {
+                    token,
+                }))
+            })?;
+
+        if current_allowance < requested_decrease {
+            return Err(Error::SafeErc20FailedDecreaseAllowance(
+                SafeErc20FailedDecreaseAllowance {
+                    spender,
+                    currentAllowance: current_allowance,
+                    requestedDecrease: requested_decrease,
+                },
+            ));
+        }
+
+        self.force_approve(
+            token,
+            spender,
+            current_allowance - requested_decrease,
+        )
+    }
+
+    /// Set the calling contract's allowance toward `spender` to `value`. If
+    /// `token` returns no value, non-reverting calls are assumed to be
+    /// successful. Meant to be used with tokens that require the approval
+    /// to be set to zero before setting it to a non-zero value, such as USDT.
+    pub fn force_approve(
+        &mut self,
+        token: Address,
+        spender: Address,
+        value: U256,
+    ) -> Result<(), Error> {
+        let selector = function_selector!("approve", Address, U256);
+
+        // Helper function to construct calldata
+        fn build_approve_calldata(
+            spender: Address,
+            value: U256,
+            selector: &[u8],
+        ) -> Vec<u8> {
+            let encoded_args = (spender, value).abi_encode_params();
+            [&selector[..4], &encoded_args].concat()
+        }
+
+        // Try performing the approval with the desired value
+        let approve_data = build_approve_calldata(spender, value, &selector);
+        if self._call_optional_return(token, &approve_data).is_ok() {
+            return Ok(());
+        }
+
+        // If that fails, reset allowance to zero, then retry the desired
+        // approval
+        let reset_data = build_approve_calldata(spender, U256::ZERO, &selector);
+        self._call_optional_return(token, &reset_data)?;
+        self._call_optional_return(token, &approve_data)?;
+
+        Ok(())
+    }
+}
+
+impl SafeErc20 {
+    /// Imitates a Stylus high-level call, relaxing the requirement on the
+    /// return value: if data is returned, it must not be `false`, otherwise
+    /// calls are assumed to be successful.
+    fn _call_optional_return(
+        &self,
+        token: Address,
+        data: &[u8],
+    ) -> Result<(), Error> {
+        match RawCall::new()
+            .gas(gas_left())
+            .limit_return_data(0, 32)
+            .call(token, data)
+        {
+            Ok(data)
+                if (data.is_empty() && Address::has_code(&token))
+                    || (!data.is_empty() && encodes_true(&data)) =>
+            {
+                Ok(())
+            }
+            _ => {
+                Err(Error::SafeErc20FailedOperation(SafeErc20FailedOperation {
```

Comment:
I would unpack data with `.map_err(|_| Error::SafeErc20FailedOperation(SafeErc20FailedOperation )?` and then perform the verification later.